contracts/ClaimAndStake.sol

```diff
@@ -0,0 +1,81 @@
+}
+
+/**
+ * @notice Allows claimer to claim tokens from AcrossMerkleDistributor and stake into AcceleratingDistributor
+ * atomically in a single transaction. This intermediary contract also removes the need for claimer to approve
+ * AcceleratingDistributor to spend its staking tokens.
+ */
+
+contract ClaimAndStake is ReentrancyGuard, Ownable, Multicall {
+    using SafeERC20 for IERC20;
+
+    // Contract which rewards tokens to users that they can then stake.
+    AcrossMerkleDistributorInterface public merkleDistributor;
+
+    // Contract that user stakes claimed tokens into.
+    AcceleratingDistributorInterface public acceleratingDistributor;
+
+    /**************************************
+     *               EVENTS               *
+     **************************************/
+    event SetMerkleDistributor(AcrossMerkleDistributorInterface indexed newMerkleDistributor);
+    event SetAcceleratingDistributor(AcceleratingDistributorInterface indexed newAcceleratingDistributor);
+
+    constructor() {}
+
+    /**************************************
+     *          ADMIN FUNCTIONS           *
+     **************************************/
+
+    /**
+     * @notice Sets merkle and accelerating distributor contracts called in claimAndStake.
+     * @param _merkleDistributor Address to set merkleDistributor to.
+     * @param _acceleratingDistributor Address to set acceleratingDistributor to.
+     */
+    function setDistributorContracts(
+        AcrossMerkleDistributorInterface _merkleDistributor,
+        AcceleratingDistributorInterface _acceleratingDistributor
+    ) external nonReentrant onlyOwner {
+        merkleDistributor = _merkleDistributor;
+        acceleratingDistributor = _acceleratingDistributor;
+        emit SetMerkleDistributor(_merkleDistributor);
+        emit SetAcceleratingDistributor(_acceleratingDistributor);
+    }
+
+    /**
+     * @notice Claim tokens from a MerkleDistributor contract and stake them for rewards in AcceleratingDistributor.
+     * @dev Will revert if `merkleDistributor` is not set to valid MerkleDistributor contract.
+     * @dev Will revert if the claim recipient account is not equal to caller, or if the reward token
+     *      for claim is not a valid staking token.
+     * @dev Will revert if this contract is not a "whitelisted claimer" on the MerkleDistributor contract.
+     * @param _claim Claim leaf to retrieve from MerkleDistributor.
+     * @param stakedToken The address of the token to stake.
+     */
+    function claimAndStake(MerkleDistributorInterface.Claim memory _claim, address stakedToken) external nonReentrant {
+        require(_claim.account == msg.sender, "claim account not caller");
+        require(merkleDistributor.getRewardTokenForWindow(_claim.windowIndex) == stakedToken, "unexpected claim token");
+        merkleDistributor.claimFor(_claim);
+        IERC20(stakedToken).safeApprove(address(acceleratingDistributor), _claim.amount);
```

Comment:
I wanted to get your thoughts on the use of this function. We usually prefer to _not_ use it, but it might actually be _good_ to use it in this case.

See [this check](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/d5ca39e9a264ddcadd5742484b6d391ae1647a10/contracts/token/ERC20/utils/SafeERC20.sol#L54-L57).

I was originally thinking that it would be bad if a leftover approval from a previous call caused the entire contract to begin reverting. However, maybe the opposite is true. A leftover nonzero approval might mean that there is a bug in the contract, so it could be preferable for it to begin reverting in that case.

This probably doesn't matter much, but the alternative would be [this function](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/d5ca39e9a264ddcadd5742484b6d391ae1647a10/contracts/token/ERC20/utils/SafeERC20.sol#L61) if we wanted to _not fail_ if there was leftover approval. Thoughts?

Reply:
Oh good point this was an oversight i think its always safer to use safeIncreaseAllowance